### PR TITLE
MPP-3636: Emit glean pings as logs

### PR DESCRIPTION
**Pull request by jwhitlock**
For MPP-3636:

* Add `glean_parser` to requirements
* Add a prototype Glean metrics YAML
* Add prototype interface code to emit a ping as a log
* Add metrics pings for mask creation events:
  - Random mask created via the API
  - Domain mask created via the API
  - Domain mask created via an incoming email

This has related changes that could be moved to their own PR(s):

* Wrap code and comments that exceed 88 columns, making `flake8` happy in files touched by this PR
* Force `settings.RELAY_CHANNEL` to be one of "local", "dev", "stage", "prod". Previously, other values could be set in the environment.
* Fix an issue where recreating a domain mask raised an exception with a nonsensical message. This code now has tests.
* Refactor loading version information from `version.json`, created by CircleCI, so the version number can be used in the Glean code as well. This code is used by the `/__version__` view, and now has tests.
* Add type hints to `emails.models.Profile.fxa`, to say that it can return `None` or the user's `SocialAccount`. Add checks and assertions for the code that uses this property to check that the value is not `None`.
* Generate UUIDs for FxA IDs in tests, rather than very long random strings, to more closely match read FxA IDs.

## Glean Integration

This PR does not send data to Glean. That requires [several steps](https://mozilla.github.io/glean/book/user/adding-glean-to-your-project/index.html):

* Complete the glean metrics YAML (in progress)
* Add a `python_server` outputter to `glean_parser` (using this PR as a guide)
* Add `glean_parser` to the build system
* Go through the [data review process](https://wiki.mozilla.org/Data_Collection)
* Add Relay to the probe scraper
* Forward glean logs to the Glean backend injestion system

This PR is similar to the initial Glean integration PRs on other projects:

* https://github.com/mozilla/fxa/pull/15659 (Node.js)
* https://github.com/MozillaSocial/mastodon/pull/36 (Ruby on Rails)

The prototype glean metrics YAML is at `telemetry/glean/relay-server-metrics.yaml`. This directory structure allows for future front-end metrics YAML, as well as other telemetry files for logs, metrics, etc. It has a single event, `email.generate_mask`, for mask generation and related data. The complete file will have many other events, and this event may change significantly.

The prototype interface file is at `privaterelay/glean/server_events.py`. It is a Python translation of the `javascript_server` format. First, I generated the JS version:

```sh
glean_parser translate -f javascript_server -o privaterelay/glean telemetry/glean/relay-server-metrics.yaml
```

I then hand-translated the JavaScript `server_events.js` into similar Python `server_events.py`. When `glean_parser` supports `python_server`, we'll use that instead.

## Glean log format

The Glean log entry follows the lead of other projects sending Glean pings as log entries. A Glean log entry, identified by `"Type": "glean"`, looks like:

```json
{"Timestamp": 1705616776198056960, "Type": "glean", "Logger": "fx-private-relay", "Hostname": "jwhitlock-MB-M2.lan", "EnvVersion": "2.0", "Severity": 6, "Pid": 26448, "Fields": {"document_namespace": "relay-backend", "document_type": "events", "document_version": "1", "document_id": "8109c6f3-1f91-469d-ade4-9dfc0ce258c7", "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:123.0) Gecko/20100101 Firefox/123.0", "ip_address": "", "payload": "{\"metrics\": {}, \"events\": [{\"category\": \"email\", \"name\": \"generate_mask\", \"extra\": {\"mozilla_accounts_id\": \"353d4f404d15425892dcf320d8b836ff\", \"is_random_mask\": true, \"created_by_api\": true, \"has_generated_for\": false}, \"timestamp\": 1705616776198}], \"ping_info\": {\"seq\": 0, \"start_time\": \"2024-01-18T22:26:16.198028+00:00\", \"end_time\": \"2024-01-18T22:26:16.198028+00:00\"}, \"client_info\": {\"telemetry_sdk_build\": \"glean_parser v11.0.1\", \"first_run_date\": \"Unknown\", \"os\": \"Unknown\", \"os_version\": \"Unknown\", \"architecture\": \"Unknown\", \"app_build\": \"Unknown\", \"app_display_version\": \"unknown\", \"app_channel\": \"local\"}}", "msg": "glean-server-event"}}
```

As pretty-printed JSON:

```json
{
  "Timestamp": 1705616776198056960,
  "Type": "glean",
  "Logger": "fx-private-relay",
  "Hostname": "jwhitlock-MB-M2.lan",
  "EnvVersion": "2.0",
  "Severity": 6,
  "Pid": 26448,
  "Fields": {
    "document_namespace": "relay-backend",
    "document_type": "events",
    "document_version": "1",
    "document_id": "8109c6f3-1f91-469d-ade4-9dfc0ce258c7",
    "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:123.0) Gecko/20100101 Firefox/123.0",
    "ip_address": "",
    "payload": "{\"metrics\": {}, \"events\": [{\"category\": \"email\", \"name\": \"generate_mask\", \"extra\": {\"mozilla_accounts_id\": \"353d4f404d15425892dcf320d8b836ff\", \"is_random_mask\": true, \"created_by_api\": true, \"has_generated_for\": false}, \"timestamp\": 1705616776198}], \"ping_info\": {\"seq\": 0, \"start_time\": \"2024-01-18T22:26:16.198028+00:00\", \"end_time\": \"2024-01-18T22:26:16.198028+00:00\"}, \"client_info\": {\"telemetry_sdk_build\": \"glean_parser v11.0.1\", \"first_run_date\": \"Unknown\", \"os\": \"Unknown\", \"os_version\": \"Unknown\", \"architecture\": \"Unknown\", \"app_build\": \"Unknown\", \"app_display_version\": \"unknown\", \"app_channel\": \"local\"}}",
    "msg": "glean-server-event"
  }
}
```

The `Fields.payload` entry is a JSON-encoded string, which decodes as:

```json
{
  "metrics": {},
  "events": [
    {
      "category": "email",
      "name": "generate_mask",
      "extra": {
        "mozilla_accounts_id": "353d4f404d15425892dcf320d8b836ff",
        "is_random_mask": true,
        "created_by_api": true,
        "has_generated_for": false
      },
      "timestamp": 1705616776198
    }
  ],
  "ping_info": {
    "seq": 0,
    "start_time": "2024-01-18T22:26:16.198028+00:00",
    "end_time": "2024-01-18T22:26:16.198028+00:00"
  },
  "client_info": {
    "telemetry_sdk_build": "glean_parser v11.0.1",
    "first_run_date": "Unknown",
    "os": "Unknown",
    "os_version": "Unknown",
    "architecture": "Unknown",
    "app_build": "Unknown",
    "app_display_version": "unknown",
    "app_channel": "local"
  }
}
```

### CODE REVIEW

File: privaterelay/utils.py
Comment (groovecoder) [inline privaterelay/utils.py:513]: praise: ooo, nice helper! I want to use this to include the version in the google cloud profiler data too! (https://github.com/mozilla/fx-private-relay/pull/4321/files#diff-26ec47835bb87054210dfba58b8590e7d2bb34d187de1d791d9d8c9c00a7baafR24-R31)

File: privaterelay/glean/server_events.py
Comment (groovecoder) [inline privaterelay/glean/server_events.py:8]: question (non-blocking): should we add a python_server format to glean_parser up-front? I.e., should the glean pings block and wait on a python server formatter? I'd hate to merge all this good code if it's going to have to be manually re-converted all over again?

Comment (jwhitlock) [reply privaterelay/glean/server_events.py:8]: Short answer: no, we should not wait on `glean_parser` to add a `python_server` format. Let me know if you want a longer answer.

File: api/views/__init__.py
Comment (groovecoder) [inline api/views/__init__.py:204]: same question (non-blocking): in API `perform_create` or in a `post_save` signal handler?

File: emails/tests/views_tests.py
Comment (groovecoder) [inline emails/tests/views_tests.py:501]: praise: TIL `assertLogs` and `assertNoLogs`. question (non-blocking): when do we use `assertLogs` vs. a pytest fixture?

File: privaterelay/views.py
Comment (groovecoder) [inline privaterelay/views.py:37]: praise: nice tiny clean-up.

File: telemetry/glean/relay-server-metrics.yaml
Comment (groovecoder) [inline telemetry/glean/relay-server-metrics.yaml:19]: question (blocking): does these need real values before merge?

**Review by groovecoder — CHANGES_REQUESTED** [review]
Everything looks really good.

I tried to see what would happen if I added an undefined argument to `record_email_generate_mask`. The dashboard was still able to create a new mask, but it also showed an error toast:
image

It looks like `mypy` caught my bad code anyway:
```
api/views/__init__.py:135: error: Unexpected keyword argument "some_other_field" for "record_email_generate_mask" of "EventsServerEventLogger"  [call-arg]
privaterelay/glean/server_events.py:82: note: "record_email_generate_mask" of "EventsServerEventLogger" defined here
api/views/__init__.py:135: error: Unexpected keyword argument "some_other_field" for "record_email_generate_mask" of "EventsServerEventLogger"  [call-arg]
```

But I wonder: should we have a specific CI job check just on the glean pings?

File: api/views/__init__.py
Comment (jwhitlock) [reply api/views/__init__.py:204]: The `post_save` signal doesn't have access to the request headers

File: emails/tests/views_tests.py
Comment (jwhitlock) [reply emails/tests/views_tests.py:501]: `caplog` when the test is a function, `assertLogs` when it is a method in a `unittest.TestCase` class

File: telemetry/glean/relay-server-metrics.yaml
Comment (jwhitlock) [reply telemetry/glean/relay-server-metrics.yaml:19]: I'm coming around to not merging this PR - maybe this is just a spike.

**Comment by groovecoder** [comment]
Should we close this PR? Obviously keep the branch around, but it looks like you've effectively split out the merge-able changes into their own PRs.

**Comment by jwhitlock** [comment]
Yep, let's close, this code will come back in a few weeks better and stronger

**Pull request closed without merging**
